File: Problems/Dynamic-programming/House-Robber/House-Robber.cs
using System;

// Using bottom-up approach - Time: O(n)

class Solution
{
    public int Rob(int[] nums)
    {
        int rob = 0, skip = 0;
        
        foreach (int n in nums)
        {
            int r = n + skip;
            int s = Math.Max(rob, skip);
            rob = r;
            skip = s;
        }

        return Math.Max(rob, skip);
    }
}

class Program
{
    static void Main(string[] args)
    {
        int[] nums = { 2, 7, 9, 3, 1 };
        Console.WriteLine("Input: nums = [" + string.Join(", ", nums) + "]");

        Solution sol = new Solution();
        int result = sol.Rob(nums);
        
        Console.WriteLine("Output: " + result);
    }
}

File: Problems/Dynamic-programming/House-Robber/House-Robber.cpp
#include <iostream>
#include <vector>
#include <algorithm>

// Using bottom-up approach - Time: O(n)

class Solution {
public:
    int rob(std::vector<int>& nums) 
    {
        int rob = 0, skip = 0;
    
        for (int n : nums) 
        {
            int r = n + skip, s = std::max(rob, skip);
            rob = r, skip = s;
        }
        
        return std::max(rob, skip);
    }
};

int main()
{
    std::vector<int> nums = { 1, 2, 3, 1 };
    std::cout << "Input: nums = [";
    for (int i = 0; i < nums.size(); i++)
    {
        std::cout << nums[i];
        if (i != nums.size() - 1)
            std::cout << ", ";
    }
    std::cout << "]" << std::endl;

    Solution sol;
    int result = sol.rob(nums);
    
    std::cout << "Output: "<< result << std::endl;

    return 0;
}
